File: core/streaming/dom_mutation_stream_engine.py
from __future__ import annotations

from typing import Any, Dict, List

from core.crypto.kaalka_hash_engine import compute_kaalka_hash
from core.streaming.stream_capture_engine import make_stream_event

MAX_MUTATIONS = 10000


def capture_dom_mutations(page: Any) -> Dict[str, Any]:
    mutations: List[Dict[str, Any]] = []
    events: List[Dict[str, Any]] = []

    if page is not None and hasattr(page, "_test_dom_mutations"):
        mutations = list(page._test_dom_mutations)[:MAX_MUTATIONS]

    for index, mutation in enumerate(mutations):
        events.append(
            make_stream_event(
                step=index,
                source="dom_mutation",
                direction=str(mutation.get("type", "mutation")),
                payload=str(mutation.get("payload", "")),
                connection_id=str(mutation.get("node_id", "")),
            )
        )

    dom_snapshot = ""
    if page is not None and hasattr(page, "_test_html"):
        dom_snapshot = str(page._test_html)

    return {
        "mutations": mutations,
        "events": events,
        "dom_hash": compute_kaalka_hash(dom_snapshot[:1_000_000]),
        "bounded": True,
    }

File: src/streaming/domMutationStreamEngine.ts
/**
 * Converted from Python: core/streaming/dom_mutation_stream_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { computeKaalkaHash } from "../crypto/kaalkaHashEngine.js";
import { makeStreamEvent } from "./streamCaptureEngine.js";

export let MAX_MUTATIONS: any = 10000;
export function captureDomMutations(page: any): any {
  var mutations: any[] = [];
  var events: any[] = [];
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("_test_dom_mutations") in (page as object) || typeof (page as Record<string, unknown>)[String("_test_dom_mutations")] === "function")))) {
    mutations = py.slice([...py.iter(page._test_dom_mutations)], null, MAX_MUTATIONS);
  }
  var index: any;
  var mutation: any;
  for ([index, mutation] of py.enumerate(mutations)) {
    py.listAppend(events, makeStreamEvent(index, "dom_mutation", py.toStr(py.get(mutation, "type", "mutation")), py.toStr(py.get(mutation, "payload", "")), py.toStr(py.get(mutation, "node_id", ""))));
  }
  var dom_snapshot: any = "";
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("_test_html") in (page as object) || typeof (page as Record<string, unknown>)[String("_test_html")] === "function")))) {
    dom_snapshot = py.toStr(page._test_html);
  }
  return {"mutations": mutations, "events": events, "dom_hash": computeKaalkaHash(py.slice(dom_snapshot, null, 1000000)), "bounded": true};
}
export { computeKaalkaHash, makeStreamEvent };
